edgelet/edgelet-kube/src/runtime.rs

```diff
@@ -14,140 +15,237 @@ use edgelet_core::{
 
-    fn pull(&self, _: &Self::Config) -> Self::PullFuture {
-        Box::new(future::ok(()))
+    fn pull(&self, config: &Self::Config) -> Self::PullFuture {
+        // Find and generate image pull secrets.
+        if let Some(auth) = config.auth() {
+            // Have authorization for this module spec, create this if it doesn't exist.
+            let fut = auth_to_image_pull_secret(self.namespace(), auth)
+                .map_err(Error::from)
+                .map(|(secret_name, pull_secret)| {
+                    let client_copy = self.client.clone();
+                    let namespace_copy = self.namespace().to_owned();
+                    let fut = self
+                        .client
+                        .lock()
+                        .expect("Unexpected lock error")
+                        .borrow_mut()
+                        .list_secrets(self.namespace(), Some(secret_name.as_str()))
+                        .map_err(Error::from)
+                        .and_then(move |secrets| {
+                            if let Some(current_secret) = secrets.items.into_iter().find(|secret| {
+                                secret.metadata.as_ref().map_or(false, |meta| {
+                                    meta.name.as_ref().map_or(false, |n| *n == secret_name)
+                                })
+                            }) {
+                                if current_secret == pull_secret {
+                                    Either::A(Either::A(future::ok(())))
+                                } else {
+                                    let f = client_copy
+                                        .lock()
+                                        .expect("Unexpected lock error")
+                                        .borrow_mut()
+                                        .replace_secret(
+                                            secret_name.as_str(),
+                                            namespace_copy.as_str(),
+                                            &pull_secret,
+                                        )
+                                        .map_err(Error::from)
+                                        .map(|_| ());
+
+                                    Either::A(Either::B(f))
+                                }
+                            } else {
+                                let f = client_copy
+                                    .lock()
+                                    .expect("Unexpected lock error")
+                                    .borrow_mut()
+                                    .create_secret(namespace_copy.as_str(), &pull_secret)
+                                    .map_err(Error::from)
+                                    .map(|_| ());
+
+                                Either::B(f)
+                            }
+                        });
+                    Either::A(fut)
+                })
+                .unwrap_or_else(|err| Either::B(future::err(err)));
```

Comment:
Learnt recently from Arnav that if you have a case where you have either a `FutureResult` or some other kind of `Future` to return then `Either` is not strictly necessary. For example, here on line 241 we can simply rturn `fut` and on line 243 we can convert the `Result` into a `Future` via `into_future` which could produce a `FutureResult` that yielded another `Future` (on success) which can then be _flattened_ via the `flatten` function. For example:
```rust
let fut = auth_to_image_pull_secret(self.namespace(), auth)
  .map_err(Error::from)
  .map(|(secret_name, pull_secret)| {
      let client_copy = self.client.clone();
      let namespace_copy = self.namespace().to_owned();
      self
          .client
          .lock()
          .expect("Unexpected lock error")
          .borrow_mut()
          .list_secrets(self.namespace(), Some(secret_name.as_str()))
          .map_err(Error::from)
          .and_then(move |secrets| {
              if let Some(current_secret) = secrets.items.into_iter().find(|secret| {
                  secret.metadata.as_ref().map_or(false, |meta| {
                      meta.name.as_ref().map_or(false, |n| *n == secret_name)
                  })
              }) {
                  if current_secret == pull_secret {
                      Either::A(Either::A(future::ok(())))
                  } else {
                      let f = client_copy
                          .lock()
                          .expect("Unexpected lock error")
                          .borrow_mut()
                          .replace_secret(
                              secret_name.as_str(),
                              namespace_copy.as_str(),
                              &pull_secret,
                          )
                          .map_err(Error::from)
                          .map(|_| ());

                      Either::A(Either::B(f))
                  }
              } else {
                  let f = client_copy
                      .lock()
                      .expect("Unexpected lock error")
                      .borrow_mut()
                      .create_secret(namespace_copy.as_str(), &pull_secret)
                      .map_err(Error::from)
                      .map(|_| ());

                  Either::B(f)
              }
          })
  })
  .into_future()
  .flatten();
```

Reply:
That looks a lot better, done!